ggshield/scan/scannable.py

```diff
@@ -258,35 +262,56 @@ def scan(
         matches_ignore: Iterable[IgnoredMatch],
         scan_context: ScanContext,
         ignored_detectors: Optional[Set[str]] = None,
-        on_file_chunk_scanned: Callable[
-            [List[Dict[str, Any]]], None
-        ] = lambda chunk: None,
+        on_file_chunk_scanned: Callable[[List[File]], None] = lambda chunk: None,
     ) -> Results:
         logger.debug("self=%s command_id=%s", self, scan_context.command_id)
         cache.purge()
-        scannable_list = self.scannable_list
         results = []
         errors = []
         chunks = []
-        for i in range(0, len(scannable_list), MULTI_DOCUMENT_LIMIT):
-            chunks.append(scannable_list[i : i + MULTI_DOCUMENT_LIMIT])
+        for i in range(0, len(self.files), MULTI_DOCUMENT_LIMIT):
+            chunks.append(self.files[i : i + MULTI_DOCUMENT_LIMIT])
 
         headers = get_headers(scan_context)
 
         with concurrent.futures.ThreadPoolExecutor(
             max_workers=min(CPU_COUNT, 4), thread_name_prefix="content_scan"
         ) as executor:
-            future_to_scan = {
-                executor.submit(
+            chunks_for_futures = {}
+            for chunk in chunks:
+                prepared_chunk = []
+                skipped_chunk = []
+                for file in chunk:
+                    file.prepare()
+                    if file.document:
+                        prepared_chunk.append(file)
+                    else:
+                        skipped_chunk.append(file)
+
+                if skipped_chunk:
+                    # Report skipped_chunk as progress. If we only report prepared_chunk
+                    # then the progress bar won't reach 100%.
+                    on_file_chunk_scanned(skipped_chunk)
+
+                if not prepared_chunk:
+                    # No scannable files in this chunk, go to the next one
+                    continue
+
+                # Format `prepared_chunk` into the List[Dict[str, str]] expected by
+                # `GGClient.multi_content_scan()`
+                documents = [
+                    {"document": x.document, "filename": x.filename}
+                    for x in prepared_chunk
```

Comment:
If I understand well, if we skip files, we will end up with chunks smaller than `MULTI_DOCUMENT_LIMIT`.  
Don't you think we could iterate over files, handle skipped ones, and then create one future each time we reach the `MULTI_DOCUMENT_LIMIT`.  
It might not be that of an optim though.

Reply:
Sounds like a good idea, going to give it a try.

Reply:
I refactored `scan()` to do this, but the method was getting too long to my taste, so I went a bit overboard and introduced a `Scanner` class to implement the scan process, this allowed me to split the `scan()` method into helper functions.

Now `Files` is really short (it still has a `scan()` method though, to keep the changes contained for now).